core/src/impl/Kokkos_HostSharedPtr.hpp

```diff
@@ -0,0 +1,200 @@
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_IMPL_HOST_SHARED_PTR_HPP
+#define KOKKOS_IMPL_HOST_SHARED_PTR_HPP
+
+#include <Kokkos_Macros.hpp>
+#include <Kokkos_Atomic.hpp>
+
+#include <functional>
+
+namespace Kokkos {
+namespace Experimental {
+
+template <typename T>
+class MaybeReferenceCountedPtr {
+ public:
+  using element_type = T;
+
+ protected:
+  MaybeReferenceCountedPtr(T* element_ptr)
+      : m_element_ptr(element_ptr), m_control(nullptr) {}
+
+  template <class Deleter>
+  MaybeReferenceCountedPtr(T* element_ptr, const Deleter& deleter)
+      : m_element_ptr(element_ptr), m_control(nullptr) {
+    if (element_ptr) {
+      try {
+        m_control = new Control{deleter, 1};
+      } catch (...) {
+        deleter(element_ptr);
+        throw;
+      }
+    }
+  }
+
+ public:
+  KOKKOS_FUNCTION MaybeReferenceCountedPtr(
+      MaybeReferenceCountedPtr&& other) noexcept
+      : m_element_ptr(other.m_element_ptr), m_control(other.m_control) {
+    other.m_element_ptr = nullptr;
+    other.m_control     = nullptr;
+  }
+
+  KOKKOS_FUNCTION MaybeReferenceCountedPtr(
+      const MaybeReferenceCountedPtr& other) noexcept
+      : m_element_ptr(other.m_element_ptr), m_control(other.m_control) {
+#ifdef KOKKOS_ACTIVE_EXECUTION_MEMORY_SPACE_HOST
+    if (m_control) Kokkos::atomic_add(&(m_control->m_counter), 1);
+#endif
+  }
+
+  KOKKOS_FUNCTION MaybeReferenceCountedPtr& operator=(
+      MaybeReferenceCountedPtr&& other) noexcept {
+    if (&other != this) {
+      cleanup();
+      m_element_ptr       = other.m_element_ptr;
+      other.m_element_ptr = nullptr;
+      m_control           = other.m_control;
+      other.m_control     = nullptr;
+    }
+    return *this;
+  }
+
+  KOKKOS_FUNCTION MaybeReferenceCountedPtr& operator=(
+      const MaybeReferenceCountedPtr& other) noexcept {
+    if (&other != this) {
+      cleanup();
+      m_element_ptr = other.m_element_ptr;
+      m_control     = other.m_control;
+#ifdef KOKKOS_ACTIVE_EXECUTION_MEMORY_SPACE_HOST
+      if (is_reference_counted())
+        Kokkos::atomic_add(&(m_control->m_counter), 1);
+#endif
+    }
+    return *this;
+  }
+
+  KOKKOS_FUNCTION ~MaybeReferenceCountedPtr() { cleanup(); }
+
+  KOKKOS_FUNCTION T* get() const noexcept { return m_element_ptr; }
+  KOKKOS_FUNCTION T& operator*() const noexcept {
+    KOKKOS_EXPECTS(bool(*this));
+    return *get();
+  }
+  KOKKOS_FUNCTION T* operator->() const noexcept { return get(); }
+
+  // checks if the stored pointer is not null
+  KOKKOS_FUNCTION explicit operator bool() const noexcept {
+    return get() != nullptr;
+  }
+
+  // checks whether the MaybeReferenceCountedPtr does reference counting
+  // which implies managing the lifetime of the object
+  KOKKOS_FUNCTION bool is_reference_counted() const noexcept {
+    return m_control != nullptr;
```

Comment:
This still doesn't work properly for an empty `HostSharedPtr`; since the control block doesn't get allocated for that use case, this still incorrectly returns `false`. It needs to be set to some sentinel value in the non-reference-counted case